### PR TITLE
fix(ses): work around #2348 linenumber bug

### PR DESCRIPTION
**Pull request by erights**
Closes: https://github.com/endojs/endo/issues/2348 

Refs: https://github.com/Agoric/agoric-sdk/pull/9711 https://github.com/endojs/endo/issues/1798 https://github.com/endojs/endo/pull/1799 https://github.com/Agoric/agoric-sdk/issues/8662 https://github.com/Agoric/agoric-sdk/pull/9700

## Description

Prior to this PR, when you ran on Ava on Node a test written in TypeScript, you'd see something like the following in your stack traces.

```
  boot/test/bootstrapTests/stack-linenumbers.test.ts:1:104
```

This is because the TypeScript compiler compiles a TypeScripy file into one line
of JavaScript with a sourceMap that should map back into original
source positions. Node specifically makes use of that sourceMap
to produce original line-numbers. However, Node does this in a way
that resists virtualization, so the normal SES error taming cannot use
this sourceMap info.

By default, this PR does not change this behavior. However it recognizes a new `SUPPRESS_NODE_ERROR_TAMING` environment variable.

With the `SUPPRESS_NODE_ERROR_TAMING` environment variable absent
or set to `'disabled'`, you should still see stack traces as shown above

However, if you also set  the `SUPPRESS_NODE_ERROR_TAMING` environment
variable `'enabled'`, for example by doing

```sh
$ export SUPPRESS_NODE_ERROR_TAMING=enabled
```
at a bash shell, then when you run this test you should instead see
something like
```
boot/test/bootstrapTests/stack-linenumbers.test.ts:40:32
```

At https://github.com/Agoric/agoric-sdk/pull/9711 I both 
- turn this PR into an agoric-sdk patch of endo, in order to emulate this fix until the next endo-release-agoric-sdk-sync cycle, and
- add a test case that emits an error stack trace from an Ava test case written in TypeScript, to test that it works.

### Security Considerations

This new behavior only applies when `errorTaming: 'unsafe'`, on v8, and with this new environment variable enabled. 

Setting `errorTaming: 'unsafe'` already flags to sacrifice some security for a better debugging experience. But the loss of security is moderate enough --- mostly confidentiality rather than integrity --- that some may chose this setting for some production purposes.

The new behavior is a more severe loss of security that really should be used ***only during development***, not production, when even a severe loss of security is usually not an issue.

### Scaling Considerations

none
### Documentation Considerations

The behavior prior to this PR or without this environment variable enabled is an unpleasant debugging experience. However, developers won't know how to repair it, or even that it can be repaired, without explanation. Even then, the difficultly of discovery in a problem.

The names `SUPPRESS_NODE_ERROR_TAMING` and the settings `'enabled'` and `'disabled'` are by no means clear expressions of what this does. Reviewers, ***better names would be appreciated!***

### Testing Considerations

The point. As developers write and run tests written in TypeScript, they need to iterate with problems revealed by the tests, for which they need good line numbers, including into the test code.

When the environment variable is enabled, the new behavior broke some SES tests written specifically to test the old behavior. This would not happen under CI because the environment variable is not set by default, and so may not have been noticed. But it was revealed in local testing. To repair this, this PR also sets those tests up to set `process.env.SUPPRESS_NODE_ERROR_TAMING` to `'disabled'` before lockdown, protecting those tests from the external environment variable setting.

Awkwardly, at the moment https://github.com/Agoric/agoric-sdk/pull/9711 serves as the only test of this PR. This is because I failed to figure out how to configure things so I can run TypeScript tests under Ava, like https://github.com/Agoric/agoric-sdk/pull/9711 does. I tried cargo culting the configs that seemed relevant, but it didn't work.

Reviewers, if you let me know how to do this, I'll duplicate the test case from https://github.com/Agoric/agoric-sdk/pull/9711 here, which would be good.

### Compatibility Considerations

With the environment variable absent or disabled, there should be zero difference in behavior, so none.

In a development environment where this environment variable is enabled, some stack traces will be different. But outside of SES itself, nothing should depend on the contents of stack traces, so again none.

### Upgrade Considerations

No upgrade considerations.

Nothing BREAKING.

- [x] Update `NEWS.md` for user-facing changes.

### CODE REVIEW

File: packages/ses/test/error/tame-v8-error.test.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import test from 'ava';
 import '../../index.js';
 
-lockdown();
+lockdown({ errorTaming: 'safe' });
```
Comment (erights) [inline packages/ses/test/error/tame-v8-error.test.js:4]: The absence of this was a previously unnoticed bug, since omitting it left it up to environment variables to determine, which ruins the purpose of the test. Since the relevant env variables would not be set in CI, the bug would not be noticed there either.

**Comment by mhofman** [comment]
> The new behavior is a more severe loss of security that really should be used _**only during development**_, not production, when even a severe loss of security is usually not an issue.



> The names `SUPPRESS_NODE_ERROR_TAMING` and the settings `'enabled'` and `'disabled'` are by no means clear expressions of what this does. Reviewers, _**better names would be appreciated!**_

Why not re-use the existing `errorTaming` options, and a new value, like `extremely-unsafe` or `so-unsafe-you-will-be-fired-if-you-use-in-production` ([React background for the unaware](https://news.ycombinator.com/item?id=11447020))

**Comment by turadg** [comment]
> > The new behavior is a more severe loss of security that really should be used _**only during development**_, not production, when even a severe loss of security is usually not an issue.
> 
> > The names `SUPPRESS_NODE_ERROR_TAMING` and the settings `'enabled'` and `'disabled'` are by no means clear expressions of what this does. Reviewers, _**better names would be appreciated!**_
> 
> Why not re-use the existing `errorTaming` options, and a new value, like `extremely-unsafe` or `so-unsafe-you-will-be-fired-if-you-use-in-production` ([React background for the unaware](https://news.ycombinator.com/item?id=11447020))

I concur. This new case is a kind of `errorTaming`. And it would be quite surprising to the user to set  `SUPPRESS_NODE_ERROR_TAMING: enabled` but that not happen because they didn't also set `errorTaming: unsafe`.

As a develop I would expect that `errorTaming: unsafe` is unsafe, even in production, but if there's reason to expect 'unsafe' to be used in production then I think this new level should indicate that it should not be. "extremely-unsafe" risks making "unsafe" seem like it has some safety. "you-will-be…" is clear but over the top, imo. `development` or `debugging` suffice.

**Comment by erights** [comment]
I think I concur as well. I'll try it out and we'll see what we think.

**Comment by kriskowal** [comment]
Approach looks good. I propose `errorTaming: 'unsafe-debug'`.

File: packages/ses/NEWS.md [hunk 2]
```diff
@@ -1,5 +1,27 @@
 User-visible changes in SES:
 
+# Next release
+
+- When running Ava tests on Node written in TypeScript, the SES error taming
```
Comment (erights) [inline packages/ses/NEWS.md:5]: There's no need to mention Ava here. This should be for all TypeScript on Node on v8.

File: packages/ses/docs/reference.md [hunk 3]
```diff
@@ -179,9 +184,10 @@ below.
   </tr>
   <tr>
     <td><code>overrideTaming</code></td>
-    <td><code>'moderate'</code> (default) or <code>'min'</code></td>
+    <td><code>'moderate'</code> (default) or <code>'min'</code> or <code>'severe'</code></td>
     <td><code>'moderate'</code> moderates mitigations for legacy compatibility,<br>
-        <code>'min'</code> minimal mitigations for purely modern code</td>
+        <code>'min'</code> minimal mitigations for purely modern code, <br>
+        <code>'severe'</code> when moderate mitigations are inadequate</td>
```
Comment (erights) [inline packages/ses/docs/reference.md:190]: As a drive-by, I happened to notice that `'severe'` was accidentally omitted. Fixed.

**Comment by erights** [comment]
> Approach looks good. I propose `errorTaming: 'unsafe-debug'`.

Done! PTAL

File: packages/ses/NEWS.md [hunk 4]
```diff
@@ -4,6 +4,25 @@ User-visible changes in SES:
 
 - Adds support for module descriptors better aligned with XS.
 
+- When running Ava tests on Node written in TypeScript, the SES error taming
+  gives line-numbers into the generated JavaScript, in which all compiled code
+  is on line 1. This happens even with the normal development-time lockdown
+  options setting
```
Comment (turadg) [inline packages/ses/NEWS.md:10]:
```suggestion
- When running transpiled code on Node, the SES error taming
  gives line-numbers into the generated JavaScript, which often don't match the 
  the original lines. This happens even with the normal development-time lockdown
  options setting,
```

File: packages/ses/NEWS.md [hunk 5]
```diff
@@ -4,6 +4,25 @@ User-visible changes in SES:
...
+  ```js
+  errorTaming: 'unsafe'
+  ```
+  or setting the environment variable
+  ```sh
+  $ export LOCKDOWN_ERROR_TAMING=unsafe
+  ```
+  In addition to the normal `'safe'` and `'unsafe'` settings, this release
```
Comment (turadg) [inline packages/ses/NEWS.md:18]:
```suggestion
  To get the original line numbers, this release
```

File: packages/ses/NEWS.md [hunk 6]
```diff
@@ -4,6 +4,25 @@ User-visible changes in SES:
...
+  adds `'unsafe-debug'`. This `errorTaming: 'unsafe-debug'` setting
+  should be used ***during development only*** when you can
+  sacrifice more security for a better debugging experience, as explained at
+  [`errorTaming` Options](https://github.com/endojs/endo/blob/master/packages/ses/docs/lockdown.md#errortaming-options).
+  With this setting, when running Ava tests written in TypeScript on Node,
```
Comment (turadg) [inline packages/ses/NEWS.md:23]:
```suggestion
  With this setting, when running transpiled code on Node (e.g. tests written in TypeScript),
```

File: packages/ses/NEWS.md [hunk 7]
```diff
@@ -4,6 +4,25 @@ User-visible changes in SES:
...
+  the stacktrace line-numbers should point back into the original
+  TypeScript source, as they would on Node without SES.
```
Comment (turadg) [inline packages/ses/NEWS.md:25]:
```suggestion
  the stacktrace line-numbers point back into the original
  source, as they do on Node without SES.
```

File: packages/ses/docs/lockdown.md [hunk 8]
```diff
@@ -293,27 +293,48 @@ default error behavior is much more dangerous. The v8 `Error` constructor
...
+Node on v8.
 
 ```js
 lockdown(); // errorTaming defaults to 'safe'
 // or
 lockdown({ errorTaming: 'safe' }); // Deny unprivileged access to stacks, if possible
 // vs
 lockdown({ errorTaming: 'unsafe' }); // stacks also available by errorInstance.stack
+// vs
+lockdown({ errorTaming: 'unsafe-debug' }); // sacrifice more safety for TypeScript line numbers.
```
Comment (turadg) [inline packages/ses/docs/lockdown.md:337]:
```suggestion
lockdown({ errorTaming: 'unsafe-debug' }); // sacrifice more safety for sourcemapped line numbers.
```

File: packages/ses/docs/reference.md [hunk 9]
```diff
@@ -258,18 +264,30 @@ In JavaScript the stack is only available via `err.stack`, so some
...
+this is usually the right tradeoff. Please do not use it in production.
+
 ```js
 lockdown(); // errorTaming defaults to 'safe'
 // or
 lockdown({ errorTaming: 'safe' }); // Deny unprivileged access to stacks, if possible
 // vs
 lockdown({ errorTaming: 'unsafe' }); // Stacks also available by errorInstance.stack
+// vs
+lockdown({ errorTaming: 'unsafe-debug' }); // sacrifice more safety for TypeScript line numbers.
```
Comment (turadg) [inline packages/ses/docs/reference.md:290]:
```suggestion
lockdown({ errorTaming: 'unsafe-debug' }); // sacrifice more safety for sourcemapped line numbers.
```

**Review by turadg — APPROVED** [review]
Approving modulo copy changes to remove references to Ava and TypeScript.

File: packages/ses/test/error/tame-v8-error.test.js (hunk 1 above)
Comment (erights) [reply packages/ses/test/error/tame-v8-error.test.js:4]: Resolving.

File: packages/ses/NEWS.md (hunk 2 above)
Comment (erights) [reply packages/ses/NEWS.md]: Done

File: packages/ses/NEWS.md (hunk 4 above)
Comment (erights) [reply packages/ses/NEWS.md]: Done, thanks.

File: packages/ses/NEWS.md (hunk 5 above)
Comment (erights) [reply packages/ses/NEWS.md]: Done

File: packages/ses/NEWS.md (hunk 6 above)
Comment (erights) [reply packages/ses/NEWS.md]: Done

File: packages/ses/NEWS.md (hunk 7 above)
Comment (erights) [reply packages/ses/NEWS.md]: Done. Together, all this reads much better. Thanks!

File: packages/ses/docs/lockdown.md (hunk 8 above)
Comment (erights) [reply packages/ses/docs/lockdown.md]: Done

File: packages/ses/docs/reference.md (hunk 9 above)
Comment (erights) [reply packages/ses/docs/reference.md]: Done

**Review by kriskowal — APPROVED** [review]
